src/lib/plugin.js

```diff
@@ -304,91 +316,134 @@ module.exports = class PluginVirtual extends EventEmitter2 {
+    if (this._paychanBackend.handleIncomingClaim) {
+      try {
+        await this._paychanBackend.handleIncomingClaim(this._paychanContext, result)
+      } catch (e) {
+        debug('error handling incoming claim:', e)
+      }
+    }
   }
 
   async _handleFulfillCondition (transferId, fulfillment) {
     this._validator.validateFulfillment(fulfillment)
+    const transferInfo = await this._transfers.get(transferId)
 
-    const error = this._transfers.assertAllowedChange(transferId, 'executed')
-    if (error) {
-      await error
-      return true
+    // TODO: 'cancelled' or 'rejected'?
+    if (transferInfo.state === 'cancelled') {
+      throw new AlreadyRejectedError(transferId + ' has already been cancelled: ' +
+        JSON.stringify(transferInfo))
+    }
+
+    if (transferInfo.isIncoming) {
+      throw new Error(transferId + ' is incoming; refusing to fulfill.')
     }
 
-    this._transfers.assertOutgoing(transferId)
-    const transfer = this._transfers.get(transferId)
+    if (new Date(transferInfo.transfer.expiresAt).getTime() < Date.now()) {
+      throw new AlreadyRejectedError(transferId + ' has already expired: ' +
+        JSON.stringify(transferInfo))
+    }
 
-    this._validateFulfillment(fulfillment, transfer.executionCondition)
-    this._transfers.fulfill(transferId, fulfillment)
-    this._safeEmit('outgoing_fulfill', transfer, fulfillment)
-    await this._balance.sub(transfer.amount)
+    this._validateFulfillment(fulfillment, transferInfo.transfer.executionCondition)
+    await this._transfers.fulfill(transferId, fulfillment)
+    this._safeEmit('outgoing_fulfill', transferInfo.transfer, fulfillment)
+
+    let result = true
+    if (this._paychanBackend.createOutgoingClaim) {
+      try {
+        result = await this._paychanBackend.createOutgoingClaim(
+          this._paychanContext,
+          await this._transfers.getOutgoingFulfilled())
+      } catch (e) {
+        debug('error creating outgoing claim:', e)
+      }
+    }
 
-    return true
+    return result
   }
 
   async rejectIncomingTransfer (transferId, reason) {
-    const transfer = this._transfers.get(transferId)
     debug('going to reject ' + transferId)
+    const transferInfo = await this._transfers.get(transferId)
 
-    const error = this._transfers.assertAllowedChange(transferId, 'cancelled')
-    if (error) {
-      await error
-      // send another notification to our peer if the error wasn't thrown
-      await this._rpc.call('reject_incoming_transfer', this._prefix, [transferId, reason])
-      return
+    if (transferInfo.state === 'fulfilled') {
+      throw new AlreadyFulfilledError(transferId + ' has already been fulfilled: ' +
+        JSON.stringify(transferInfo))
     }
 
-    this._transfers.assertIncoming(transferId)
+    if (!transferInfo.isIncoming) {
+      throw new Error(transferId + ' is outgoing; cannot reject.')
+    }
 
+    // TODO: add rejectionReason to interface
```

Comment:
yeah, the one used to interface with the state. It's unclear whether there's a benefit to being able to store rejectionReasons, but I think we should decide